### ISSUE TITLE
Timeout while waiting for response

### ISSUE THREAD

**Issue by WoodsterDK** [i8beef/HarmonyHub#5]
Hi there,

now I can transmit commands, but it seems quite often to give a "Timeout while waiting for response"... any clue?

Furthermore is it possible to detect a single keypress on the physical remote - is that being transmitted as an event?

**Comment by i8beef** [comment]
You can attach to the MessageReceived event to do something whenever a new message is received, but it's up to you to process it as it's going to be app specific how you handle the event. I don't use that very much except for debugging though. See the test program for a simple example of attaching, and then you can play with what you want to do with messages, but it'll probably require a bit of messing around with a cached config element to make it do anything meaningful.

As for the timeouts, I'm not sure. By default the timeout on any of the RequestResponseAsync calls is 2000... but it's hard coded. Do you know which specific calls are doing this for you? Pings? GetConfigs?

**Comment by WoodsterDK** [comment]
So far I have only tried the test program. So the timeout appears when pressing and performing a c command.

The test app already subscribes to the messagereceived event, but nothing is received when pressing a button on the remote.
Can you detect button presses?

**Comment by i8beef** [comment]
When a message is published to the feed by the Harmony, the library grabs it and pipes it right out to the MessageReceived, but it will only do so for "expected" messages (iq, message, and presense). If there's anything sent over the line besides those, it'll throw an exception... so if you aren't seeing a bunch of "Unexpected XML element" exceptions, I'm going to posit that the Harmony doesn't send notifications for individual button presses... but I'm not sure on that, you'd have to stick a debugger in and watch the stream. Again, I don't do this personally, so I don't know for sure what the Harmony does. There is no published spec, several library authors, myself included, have just sniffed the lines to determine what we can about their protocol.

The timeouts I don't know either. I don't see that on mine personally. Is your network all local here? I can port in a config setting for setting a longer timeout to see if that's the issue... you aren't by chance sitting inside a debugger breakpoint when you are seeing those are you?

**Comment by WoodsterDK** [comment]
Breakpoint.... must admit I cannot remember... might be the cause.
Can check in a few days, as Im away from my hardware.

Regarding the button events, thought it was possible as it would be really nice to combine the physical remote with some actions.

**Comment by i8beef** [comment]
Well, if you are sitting in a breakpoint in between sending the request and receiving the response... that's what you are seeing.  Nevertheless, I'll add a configurable timeout value here before I release another version, it's a simple ask, and always better than hard coded values that aren't defined by spec.

As for the event emitting, my point with us reverse engineering the protocol is this: this isn't something meant to be useful to others, it's something meant to drive the phone / advanced remote interfaces. Therefore, it supports the things those need, allowing a client to send commands, or notifying a client of certain changes that they need to reflect in a specific UI, like activity change, etc. If you think about it, one remote doesn't really need to know when another remote presses a button... at least not for the interface Logitech exposes. So I wouldn't be surprised if it actually doesn't bother emitting an event on button press... 

Like I said though, this isn't really something I I've tried. IF it emits something, you WILL see it in that message event handler though, or it will throw an exception if its an unknown message type. If you aren't seeing anything, it isn't sending anything unfortunately.

**Comment by WoodsterDK** [comment]
OK - will give it a shot.
It very much looks like they don't transmit button presses as events.
Could have been nice though.... But then a custom activity for each action could be a solution.
Will see what I can come up with....

**Comment by i8beef** [comment]
Cool. I did push a new version that allows you to set a different command timeout if you want. It doesn't let you set a different "fire and forget" timeout, but that should be unnecessary I think, given the nature of the commands that go through that, and the fact that the timeout on that is just to clean up an async handler.

Ill leave this open for a little bit in case you have other questions, but I think we've covered all your questions.
